pkg/kubectl/cmd/macro/secret.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package macro
+
+import (
+	"io"
+
+	"github.com/spf13/cobra"
+
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+)
+
+// NewCmdCreateSecret groups subcommands to create various types of secrets
+func NewCmdCreateSecret(f *cmdutil.Factory, cmdOut io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "secret",
+		Short: "Create a secret using specified subcommand.",
+		Long:  "Create a secret using specified subcommand.",
+		Run: func(cmd *cobra.Command, args []string) {
+			cmd.Help()
+		},
+	}
+	cmd.AddCommand(NewCmdCreateSecretDockerRegistry(f, cmdOut))
+	cmd.AddCommand(NewCmdCreateSecretOpaque(f, cmdOut))
+	return cmd
+}
+
+const (
+	secretLong = `
+Create a secret based on a file, directory, or specified literal value
+
+A single secret may package one or more key/value pairs.  
+
+When creating a secret based on a file, the key will default to the filename, and the value will
+default to the file content.  If the filename is an invalid key, you may specify an alternate key.
```

Comment:
Is there any reason not to loosen the allowed secret keys to anything that is a valid filename?  @pmorie  It's sort of ugly and assumes something about the eventual usage of it, but I don't see the harm (though we should block leading '..' as reserved, as we do in downward API).

Or maybe we should transform filenames into valid keys?

Or maybe we should just not offer this feature (implied keys)?  That breaks dirs.